GRPO implementation

abaheti95 opened the pull request:
Adds critic free model pathway to online RL code (for [GRPO](https://arxiv.org/abs/2402.03300)) and modifies some function signatures and code pathways.
- new `hf_critic_free_lm` model class similar to DPO HF model (i.e. doesn't have `config_overrides` but directly passes additional params via model_config)
- Includes `normalize_advantage` and `length_normalize_policy_loss` flags for critic-free model. Set these both to `false` for [Dr. GRPO objective](https://arxiv.org/abs/2503.20783). The `length_normalize_policy_loss` also modifies KL length normalization.
- `ppo_loss` --> `online_rl_loss`. Shared pathway for both ppo and grpo with `critic_free` boolean flag
- `composer_ppo_forward` --> `composer_online_rl_forward`. Shared pathway for both ppo and grpo with `critic_free` path
- Logs both `seq_kl` and `token_kl` of all variants in the loss logs

Working runs with new patch:
- GRPO `mcli logs test-critic-free-pr-grpo-kl-fix-CcXvGF -f`
- PPO `mcli logs test-critic-free-pr-ppo-kl-fix-sXRGBT -f`

NOTE: Need to modify the PPO model definition to nuke `config_overrides` in the future

Autoresumption also working with GRPO
<img width="372" alt="image" src="https://github.com/user-attachments/assets/f1ad3794-7c27-47e8-b3f2-838feb584c01" />

bcui-db commented on compose_rl/ppo/callback.py:
```diff
@@ -269,9 +263,19 @@ def env_reward(
+                                          dim=-1)
+            device_train_microbatch_values *= value_action_mask
+            partial_env_output['values'
+                              ] = device_train_microbatch_values.detach()
```
```suggestion
                              ] = device_train_microbatch_values
```

abaheti95 replied:
Cool. Removed all `detach()` from `partial_env_output` within `torch.no_grad()`

abaheti95 commented on compose_rl/ppo/callback.py:
```diff
@@ -321,19 +325,44 @@ def __init__(
         # Value used in the generalized advantage estimate calculation.
         self.lambda_gae = var_config.get('lambda_gae', 1.0)
 
+        # Other algo specific hparams
```
changed to `loss_type` as discussed offline

bcui-db commented on compose_rl/ppo/callback.py:
```diff
@@ -321,19 +324,46 @@ def __init__(
+        # Other algo specific hparams
+        # Find if we are using a critic free model or not
+        if train_config['model']['name'] == 'hf_critic_free_lm':
+            self.loss_type = 'grpo'
```
Can you make this a variable in `state.model` and we get it from there? So we can add other things like RLOO, REBEL, etc.

bcui-db commented on compose_rl/ppo/model.py:
```diff
@@ -226,3 +234,78 @@ def determine_early_stop(self):
+
+    def __init__(
+        self,
+        normalize_advantage: bool = True,
```
Can you add a docstring here explaining the variables?

bcui-db commented on compose_rl/ppo/model.py:
```diff
@@ -226,3 +234,78 @@ def determine_early_stop(self):
+        ret_val = composer_online_rl_forward(
+            batch,
+            self.model,
+            loss_type='grpo',
```
When you add the above loss_type, also modify it here

bcui-db commented on compose_rl/ppo/modeling_utils.py:
```diff
@@ -116,158 +117,270 @@ def composer_ppo_forward(
     Args:
         outputs (MutableMapping): The outputs from the forward pass.
         batch (MutableMapping): The batch to compute the loss over.
+        loss_type (str): The loss type which decides whether to use critic-free or not. Defaults to "ppo".
```
```suggestion
        loss_type (str): The loss type which decides whether to use critic-free or not. Defaults: ``ppo``.
```

bcui-db commented on compose_rl/ppo/modeling_utils.py:
```diff
@@ -116,158 +117,270 @@ def composer_ppo_forward(
+            adv_masked_mean,
+            adv_masked_var,
+        )
+    elif loss_type == 'grpo':
```
Can we do two things here: 

1. Initialize `values, v_preds, value_loss, value_clip_frac, returns_mean, returns_var, val_error` to `None` above the `ppo` conditional?
2. We should do the `loss_type` check just above and it be sth like `if loss_type` not in ["ppo", "grpo"] raise the ValueError.

bcui-db commented on compose_rl/ppo/modeling_utils.py:
```diff
@@ -116,158 +117,270 @@ def composer_ppo_forward(
+    else:
+        policy_loss = utils.masked_sum(policy_loss, batch['action_mask'])
+
+    # policy_kl_logging_dict = {
+    #     f'kl/policy_kl_{k}_estimate': v for k, v in policy_kl_dict.items()
+    # }
+    # online_ift_kl_logging_dict = {
+    #     f'kl/online_ift_kl_{k}_estimate': v
+    #     for k, v in online_ift_kl_dict.items()
+    # }
```
```suggestion
```

bcui-db reviewed (approved):
Minor comments, but overall LGTM.

abaheti95 commented:
Addressed all the comments. Also made minor changes to grpo advantages. Instead of forwarding the flat advantage tensor, it now keeps the same shape of advantages for both ppo and grpo.. In grpo, the flat advantage is copied across the action_mask instead. This reduces a few lines of code.
I re-ran the baseline PPO and GRPO runs again with this change, and both are working fine.

- `mcli logs test-critic-free-pr-grpo-kl-fix-DJ4yCU -f`
- `mcli logs test-critic-free-pr-ppo-kl-fix-5lFOaf -f`
